File: backend/services/llm.py
import os
import json
from typing import Callable


async def chat_with_tools(
    messages: list[dict],
    tools: list[dict],
    handle_tool_call: Callable,
    system: str,
    extra_context: dict | None = None,
) -> dict:
    provider = os.getenv("LLM_PROVIDER", "openai")

    if provider == "anthropic":
        return await _anthropic(messages, tools, handle_tool_call, system, extra_context)
    return await _openai(messages, tools, handle_tool_call, system, extra_context)


async def _openai(messages, tools, handle_tool_call, system, extra_context):
    from openai import AsyncOpenAI
    client = AsyncOpenAI(api_key=os.getenv("OPENAI_API_KEY"))

    response = await client.chat.completions.create(
        model="gpt-4o",
        messages=[{"role": "system", "content": system}] + messages,
        tools=tools,
        tool_choice="auto",
        temperature=0.3,
    )
    msg = response.choices[0].message

    if msg.tool_calls:
        tool_results = []
        for tc in msg.tool_calls:
            result = handle_tool_call(
                tc.function.name,
                json.loads(tc.function.arguments),
                extra_context,
            )
            tool_results.append({
                "role": "tool",
                "tool_call_id": tc.id,
                "content": json.dumps(result),
            })
        followup = await client.chat.completions.create(
            model="gpt-4o",
            messages=[{"role": "system", "content": system}] + messages + [msg] + tool_results,
            temperature=0.3,
        )
        return {"reply": followup.choices[0].message.content}

    return {"reply": msg.content}


async def _anthropic(messages, tools, handle_tool_call, system, extra_context):
    from anthropic import AsyncAnthropic
    client = AsyncAnthropic(api_key=os.getenv("ANTHROPIC_API_KEY"))

    response = await client.messages.create(
        model="claude-opus-4-5",
        max_tokens=512,
        system=system,
        messages=messages,
        tools=tools,
    )

    if response.stop_reason == "tool_use":
        tool_results = []
        for block in response.content:
            if block.type == "tool_use":
                result = handle_tool_call(block.name, block.input, extra_context)
                tool_results.append({
                    "type": "tool_result",
                    "tool_use_id": block.id,
                    "content": json.dumps(result),
                })
        followup = await client.messages.create(
            model="claude-opus-4-5",
            max_tokens=512,
            system=system,
            messages=messages
                + [{"role": "assistant", "content": response.content}]
                + [{"role": "user", "content": tool_results}],
        )
        return {"reply": followup.content[0].text}

    return {"reply": response.content[0].text}


async def simple_completion(prompt: str, max_tokens: int = 80) -> str:
    """Single-turn completion — used for AI fix generation."""
    provider = os.getenv("LLM_PROVIDER", "openai")
    if provider == "anthropic":
        from anthropic import AsyncAnthropic
        client = AsyncAnthropic(api_key=os.getenv("ANTHROPIC_API_KEY"))
        r = await client.messages.create(
            model="claude-opus-4-5",
            max_tokens=max_tokens,
            messages=[{"role": "user", "content": prompt}],
        )
        return r.content[0].text.strip()

    from openai import AsyncOpenAI
    client = AsyncOpenAI(api_key=os.getenv("OPENAI_API_KEY"))
    r = await client.chat.completions.create(
        model="gpt-4o",
        messages=[{"role": "user", "content": prompt}],
        max_tokens=max_tokens,
        temperature=0.4,
    )
    return r.choices[0].message.content.strip()

File: frontend/lib/llm.ts
/**
 * lib/llm.ts
 * Provider-agnostic LLM client.
 * Set LLM_PROVIDER=openai (default) or LLM_PROVIDER=anthropic.
 */

import "server-only";
import OpenAI from "openai";
import type { z } from "zod";

export type Message = { role: "system" | "user" | "assistant"; content: string };

const provider = process.env.LLM_PROVIDER ?? "openai";

async function callOpenAI(messages: Message[]): Promise<string> {
  const client = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
  const model = process.env.OPENAI_MODEL ?? "gpt-5.5";
  const base = { model, messages, response_format: { type: "json_object" as const } };
  try {
    const res = await client.chat.completions.create({ ...base, temperature: 0.2 });
    return res.choices[0]?.message?.content ?? "";
  } catch {
    // Some models (e.g. gpt-5.5) only allow the default temperature — retry
    // without the override, keeping JSON mode so output stays valid JSON.
    const res = await client.chat.completions.create(base);
    return res.choices[0]?.message?.content ?? "";
  }
}

async function callAnthropic(messages: Message[]): Promise<string> {
  const systemMsg = messages.find((m) => m.role === "system")?.content ?? "";
  const userMsgs = messages.filter((m) => m.role !== "system");

  const res = await fetch("https://api.anthropic.com/v1/messages", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "x-api-key": process.env.ANTHROPIC_API_KEY ?? "",
      "anthropic-version": "2023-06-01",
    },
    body: JSON.stringify({
      model: "claude-opus-4-5",
      max_tokens: 2048,
      system: systemMsg,
      messages: userMsgs.map((m) => ({ role: m.role, content: m.content })),
    }),
  });
  const json = (await res.json()) as { content: Array<{ text: string }> };
  return json.content?.[0]?.text ?? "";
}

async function callLLMRaw(messages: Message[]): Promise<string> {
  return provider === "anthropic" ? await callAnthropic(messages) : await callOpenAI(messages);
}

/**
 * Call the LLM and parse the JSON response with Zod.
 * Returns null when the response is missing or invalid (callers use deterministic fallbacks).
 */
export async function callLLMJson<T>(messages: Message[], schema: z.ZodType<T>): Promise<T | null> {
  try {
    const raw = await callLLMRaw(messages);
    const parsed: unknown = JSON.parse(raw);
    const result = schema.safeParse(parsed);
    return result.success ? result.data : null;
  } catch {
    // Any failure (missing/invalid key, unknown model, network, bad JSON) →
    // null, so callers fall back to deterministic findings instead of erroring.
    return null;
  }
}
